MapleServer2/Servers/Game/DungeonManager.cs

```diff
@@ -0,0 +1,132 @@
+        }
+
+        public int GetMapInstanceId()
+        {
+            if (RecyclableMapInstanceIds.Count > 0)
+            {
+                int mapInstanceId = RecyclableMapInstanceIds.First();
+                RecyclableMapInstanceIds.Remove(mapInstanceId);
+                return mapInstanceId;
+            }
+            return MapInstanceId++;
+
+        }
+
+        //int mapInstanceId = RecyclableMapInstanceIds[mapId].First();
+        //RecyclableMapInstanceIds[mapId].Remove(mapInstanceId);
+        //        if (RecyclableMapInstanceIds[mapId].Count == 0)
+        //        {
+        //            RecyclableMapInstanceIds.Remove(mapId);
+        //        }
+        //        return mapInstanceId;
+
+        public DungeonSession CreateDungeonSession(int dungeonId, DungeonType dungeonType)
+        {
+            int dungeonSessionId = GetUniqueSessionId();
+            int dungeonInstanceId = GetMapInstanceId();
+            DungeonSession dungeonSession = new DungeonSession(dungeonSessionId, dungeonId, dungeonInstanceId, dungeonType);
+            DungeonSessionList.Add(dungeonSessionId, dungeonSession);
+            return dungeonSession;
+        }
+
+        public void RemoveDungeonSession(int dungeonSessionId)
+        {
+            if (!DungeonSessionList.ContainsKey(dungeonSessionId))
+            {
+                return;
+            }
+            int currentDungeonSessionId = DungeonSessionList[dungeonSessionId].SessionId;
+            int currentDungeonInstanceId = DungeonSessionList[dungeonSessionId].DungeonInstanceId;
+            RecyclableSessionIds.Add(currentDungeonSessionId);
+            RecyclableMapInstanceIds.Add(currentDungeonInstanceId);
+            DungeonSessionList.Remove(dungeonSessionId);
+        }
+
+        public int GetUniqueSessionId()
+        {
+            if (RecyclableSessionIds.Count > 0)
+            {
+                int sessionId = RecyclableSessionIds.First();
+                RecyclableSessionIds.Remove(sessionId);
+                return sessionId;
+            }
+            return SessionId++;
+        }
+
+        public DungeonSession GetDungeonSession(int dungeonSessionId)
+        {
+            //set dungeon of session id to x on swich dungeon
+            //if dungeonsession exists but id != dungeonid call dungeonid change
+            //update dungeonid or handle it somewhere else
+            if (!DungeonSessionList.ContainsKey(dungeonSessionId))
+            {
+                return null;
+            }
+            return DungeonSessionList[dungeonSessionId];
+        }
+
+        public DungeonSession GetDungeonSessionByInstance(int instanceId)
+        {
+            //TODO: if no dungeonsession is found, return null
+            return DungeonSessionList.FirstOrDefault(session => session.Value.DungeonInstanceId == instanceId).Value;
+        }
+
+        public bool DungeonUsesMap(DungeonSession dungeonSession, FieldManager fieldManager, Player player)
```

Comment:
I think need function need a better name and if you have returns inside ifs, you don't need to else part, it would help to understand the function easier too!

Reply:
Changed the function name.